AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSHTTPURLResponse.swift

```diff
@@ -0,0 +1,80 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// `AWSHTTPURLResponse` contains the response body and metadata associated with the response to an HTTP request.
+///
+/// When using `AWSAPIPlugin`, you can optionally type cast `HTTPURLResponse` to an instances of `AWSHTTPURLResponse`.
+/// The response body can be accessed from the `body: Data?` property. For example, when the `APIError` is an
+/// `.httpStatusError(StatusCode, HTTPURLResponse)`, then access the response body by type casting the response to
+/// an `AWSHTTPURLResponse` and retrieve the `body` field.
+/// ```swift
+/// if case let .httpStatusError(statusCode, response) = error, let awsResponse = response as? AWSHTTPURLResponse {
+///     if let responseBody = awsResponse.body {
+///         print("Response contains a response body")
+///     }
+/// }
+/// ```
+/// **Note**: The class inheritance to `HTTPURLResponse` is to provide above mechanism, and actual
+/// implementation acts as a facade that stores an instance of `HTTPURLResponse` that delegates overidden methods to
+/// this stored property.
+public class AWSHTTPURLResponse: HTTPURLResponse {
+
+    /// The response body of the response
+    public let body: Data?
+
+    private let response: HTTPURLResponse
+
+    init?(response: HTTPURLResponse, body: Data?) {
```

Comment:
If we're subclassing, why assign to a private instance rather than invoking [`super.init(url:statusCode:httpVersion:headerFields:)`](https://developer.apple.com/documentation/foundation/httpurlresponse/1415870-init). That way we don't have to proxy all the calls below either.

Reply:
I felt that this initializer was a bit limiting (lost some information about the parent `URLResponse` class) which led me to think its purpose is meant for mocking. If we take a look at the super.init's parameters, I could avoid the proxying by calling super.init with the following values
- `url` - can directly use `response.url`
- `statusCode` - can directly use `response.statusCode`
- `httpVersion` - could not find this on the response object, could hard code to "HTTP/1.1" as a dummy value
- `headerFields` - translate from `response.allHeaderValues` of type `[AnyHashable: Any]` to [String: String]`

By doing so, the values returned are correct (header fields, status code). My concern with this was developers could also access the `URLResponse` methods from the `HTTPURLResponse` instance, which would then be missing information such as `mimeType`, `expectedContentLength`, `textEncoding`. So because I could not fully re-create an instance that had the same information as the original instance, I had decided to hold onto the original instance and proxy the calls, incase this missing information causes a problem for existing developers.

Reply:
That makes a lot of sense to me, thanks for explaining.